COMPONENT-GUIDELINES.md

```diff
@@ -0,0 +1,293 @@
+
+## Process
+
+To create and submit a new or updated component to the Pipedream registry:
+
+1. Fork or branch the `pipedreamhq/pipedream` repo
+2. Develop and test a new or updated component that conforms to the guidelines in the document
+3. Create a PR for the Pipedream team to review and post a message in our [community forum](https://pipedream.com/community/c/dev/11)
+4. Address feedback provided by Pipedream
+5. Once Pipedream completes it review, the team will merge the PR to the `master` branch
+
+Looking for ideas? Check out [sources](https://github.com/PipedreamHQ/pipedream/issues?q=is%3Aissue+is%3Aopen+%5BSOURCE%5D+in%3Atitle) and [actions](https://github.com/PipedreamHQ/pipedream/issues?q=is%3Aissue+is%3Aopen+%5BACTION%5D+in%3Atitle+) requested by the community!
+
+# Reference Components
+
+The following components may be used as models for developing sources and actions for Pipedream's registry.
+
+**Sources**
+
+| Name                                                         | App          | Type                                         |
+| ------------------------------------------------------------ | ------------ | -------------------------------------------- |
+| [New Card](components/trello/sources/new-card/new-card.js)   | Trello       | Webhook                                      |
+| [Search Mentions](components/twitter/sources/search-mentions/search-mentions.js) | Twitter      | Polling                                      |
+| [New or Modified Files](components/google_drive/sources/new-or-modified-files/new-or-modified-files.js) | Google Drive | Webhook + Polling                            |
+| [New Submission](components/jotform/sources/new-submission/new-submission.js) | Jotform      | Webhook (with no unique hook ID)             |
+| [New Stars](components/github/sources/new-star/new-star.js)  | Github       | Webhook (with extensive use of common files) |
+
+**Actions**
+
+*Coming soon*
+
+# Guidelines & Patterns
+
+## General
+
+### Component Scope
+
+Create components to address specific use cases whenever possible. For example, when a user subscribes to a Github webhook to listen for “star” activity events can be generated when users star or unstar a repository. The “New Star” source filters events for only new star activity so the user doesn’t have to.
+
+There may be cases where it's valuable to create a generic component that provides users with broad lattitude (e.g., see the [custom webhook](components/github/sources/custom-webhook-events) event source for Github). However, as a general heuristic, we found that tightly scoped components are easier for users to understand and use.
+
+### Required Metadata
+
+Registry components require a unique key and version, and a friendly name and description. Action components require a `type` field to be set to `action` (sources will require a type to be set in the future).
```

Comment:
Might be worth an example stanza to clarify inline, e.g.:

```javascript
key: "google_drive-new-shared-drive",
name: "New Shared Drive",
description: "Emits a new event any time a shared drive is created.",
version: "0.0.1",
```

`key` also isn't documented in the Component API, so might be worth adding and linking to that from here so users know how to structure the format of the key.